metadata/snapshot.go

```diff
@@ -464,10 +588,38 @@ func (s *snapshotter) Commit(ctx context.Context, name, key string, opts ...snap
 
 		inheritedOpt := snapshots.WithLabels(filterInheritedLabels(base.Labels))
 
-		// TODO: Consider doing this outside of transaction to lessen
-		// metadata lock time
-		return s.Snapshotter.Commit(ctx, nameKey, bkey, inheritedOpt)
-	})
+		// NOTE: Backend snapshotters should commit fast and reliably to
+		// prevent metadata store locking and minimizing rollbacks.
+		// This operation should be done in the transaction to minimize the
+		// risk of the committed keys becoming out of sync. If this operation
+		// succeed and the overall transaction fails then the risk of out of
+		// sync data is higher and may require manual cleanup.
+		if err := s.Snapshotter.Commit(ctx, nameKey, bkey, inheritedOpt); err != nil {
```

Comment:
> If this operation succeed and the overall transaction fails then the risk of out of sync data is higher and may require manual cleanup.

The gc will cleanup the backend data if metadata boltdb side doesn't have it, right?

Reply:
The gc protects against unreferenced data, so it is clean from that perspective. The issue that still needs to be thought more about here is about how to transition the snapshot from active to commited reliably on the backend and metadata. The easiest way to do that now is to have the commit inside the transaction. However, a backend may be performing expensive operations on commit (such as copying files or computing size) before returning, keeping that write transaction open much longer than desired. An example of the risk would be if the commit succeeds in the backend but fails in the metadata store. In that case the metadata will keep referencing a backend id which no longer exists and the gc will end up removing the committed snapshot.  We would need a "pre-commit" state in the database or something and have the GC capable of rolling back those incomplete transactions. I think this is work that we should do eventually though to help keep the database unlocked as much as possible.